crates/wasmtime/src/compile/scc.rs

```diff
@@ -0,0 +1,528 @@
+                    debug_assert!(indices[node].is_some());
+                    debug_assert!(lowlinks[node].is_some());
+                    debug_assert!(lowlinks[node] <= indices[node]);
+                    debug_assert!(indices[succ].is_some());
+                    debug_assert!(lowlinks[succ].is_some());
+                    debug_assert!(lowlinks[succ] <= indices[succ]);
+
+                    // If the successor is still on the SCC stack, then it is
+                    // part of the same SCC as this node, so propagate its
+                    // lowlink to this node's lowlink.
+                    if on_stack.contains(succ) {
+                        lowlinks[node] = Some(std::cmp::min(
+                            lowlinks[node].unwrap(),
+                            lowlinks[succ].unwrap(),
+                        ));
+                    }
+                }
+
+                DfsEvent::Post(node) => {
+                    debug_assert!(indices[node].is_some());
+                    debug_assert!(lowlinks[node].is_some());
+
+                    // If this node's index is the same as its lowlink, then it
+                    // is the root of a component. Pop this component's elements
+                    // from the SCC stack and push them into our result data
+                    // structures.
+                    if indices[node] == lowlinks[node] {
+                        let start = component_nodes.len();
+                        let start = u32::try_from(start).unwrap();
+                        loop {
+                            let v = stack.pop().unwrap();
+                            let was_on_stack = on_stack.remove(v);
+                            debug_assert!(was_on_stack);
+                            component_nodes.push(v);
+                            if v == node {
+                                break;
+                            }
+                        }
+                        let end = component_nodes.len();
+                        let end = u32::try_from(end).unwrap();
+                        debug_assert!(end > start);
+                        components.push(start..end);
+                    }
+                }
+            }
+        }
+
+        Self {
+            components,
+            component_nodes,
+        }
+    }
+
+    /// Get the number of components.
+    pub fn len(&self) -> usize {
+        self.components.len()
+    }
+
+    /// Iterate over each strongly-connnected component and the `Node`s that are
+    /// members of it.
+    ///
+    /// Iteration happens in reverse-topological order (successors are visited
+    /// before predecessors in the resulting SCC DAG).
+    pub fn iter(&self) -> impl ExactSizeIterator<Item = (Component, &[Node])> + '_ {
+        self.components.iter().map(|(component, range)| {
+            let start = usize::try_from(range.start).unwrap();
+            let end = usize::try_from(range.end).unwrap();
```

Comment:
Can we reuse `self.nodes(component)` here and below?

Reply:
That technically implies an extra bounds check unless LLVM optimizes it away, but I can factor out the range conversion and indexing bits into a shared helper.

Reply:
Ah, good point! Not a huge deal, just noticed the duplication.